plotly/tools.py

```diff
@@ -1269,3 +1277,628 @@ def return_figure_from_figure_or_data(figure_or_data, validate_figure):
+        X, Y = np.meshgrid(x, y)
+        ss = 5.0
+        x_s, y_s = -1.0, 0.0
+
+        # Compute the velocity field on the mesh grid
+        u_s = ss/(2*np.pi) * (X-x_s)/((X-x_s)**2 + (Y-y_s)**2)
+        v_s = ss/(2*np.pi) * (Y-y_s)/((X-x_s)**2 + (Y-y_s)**2)
+
+        # Create streamline
+        streamline = TraceFactory.create_streamline(x, y, u_s, v_s, density=2,
+                                                name='streamline')
+
+        # Add source point
+        point = Scatter(x=[x_s], y=[y_s], mode='markers',
+                        marker=Marker(size=14), name='source point')
+        # Plot
+        fig=Figure()
+        fig['data'].append(streamline)
+        fig['data'].append(point)
+        py.iplot(fig, filename='streamline')
+        ```
+        """
+        streamline_x, streamline_y = Streamline(x, y, u, v,
+                                                density, angle,
+                                                arrow_scale).sum_streamlines()
+        arrow_x, arrow_y = Streamline(x, y, u, v,
+                                      density, angle,
+                                      arrow_scale).get_streamline_arrows()
+
+        streamline = Scatter(x=streamline_x + arrow_x,
+                             y=streamline_y + arrow_y,
+                             mode='lines', **kwargs)
+        return streamline
+
+
+class Quiver(TraceFactory):
+    def __init__(self, x, y, u, v,
+                 scale, arrow_scale, angle, **kwargs):
+        try:
+            x = self.flatten(x)
+        except exceptions.PlotlyError:
+            pass
+
+        try:
+            y = self.flatten(y)
+        except exceptions.PlotlyError:
+            pass
+
+        try:
+            u = self.flatten(u)
+        except exceptions.PlotlyError:
+            pass
+
+        try:
+            v = self.flatten(v)
+        except exceptions.PlotlyError:
+            pass
+
+        self.x = x
+        self.y = y
+        self.u = u
+        self.v = v
+        self.scale = scale
+        self.arrow_scale = arrow_scale
+        self.angle = angle
+        self.end_x = []
+        self.end_y = []
+        self.validate()
+        self.scale_uv()
+        barb_x, barb_y = self.get_barbs()
+        arrow_x, arrow_y = self.get_quiver_arrows()
+
+    def flatten(self, array):
+        """
+        Uses list comprehension to flatten array
+
+        :param (array): An iterable to flatten
+        :raises (PlotlyError): If iterable is not nested.
+        :rtype (list): The flattened list.
+        """
+        try:
+            return [item for sublist in array for item in sublist]
+        except TypeError:
+            raise exceptions.PlotlyError("Your data array could not be"
+                                         "flattened! Make sure x, y, u, and v"
+                                         "are lists or numpy ndarrays!")
+
+    def validate(self):
+        """
+        Validates that args and kwargs meet criteria,
```

Comment:
:cow2: There are generally two ways to do docstrings.

* One-liners https://www.python.org/dev/peps/pep-0257/#one-line-docstrings
* Multi-liners https://www.python.org/dev/peps/pep-0257/#multi-line-docstrings

For a multi-line docstring, docstring parsers will return the first line expecting it to be a summary. So, this would be best written as:

```
"""
Validates the user-input arguments.

Specifically, this checks that scale and arrow_scale are positive
and that x, y, u, and v are the same length.

:raises: (ValueError) If scale or arrow_scale is < 1.
:raises: (PlotlyError) If x, y, u, and v are not the same length.

"""
```

Nbd on this one, mostly just a heads up about the *first* line convention.